packages/keyring-controller/src/KeyringController.test.ts

```diff
@@ -3213,6 +3213,39 @@ describe('KeyringController', () => {
       });
     });
   });
+
+  describe('atomic operations', () => {
+    describe('addNewKeyring', () => {
+      it('should rollback the controller keyrings if the keyring creation fails', async () => {
+        const mockAddress = '0x1234';
+        stubKeyringClassWithAccount(MockKeyring, mockAddress);
+        // Mocking the serialize method to throw an error will
+        // halt the controller everytime it tries to persist the keyring,
+        // making it impossible to update the vault
+        jest
+          .spyOn(MockKeyring.prototype, 'serialize')
+          .mockImplementation(async () => {
+            throw new Error('You will never be able to persist me!');
+          });
+        await withController(
+          { keyringBuilders: [keyringBuilderFactory(MockKeyring)] },
+          async ({ controller, initialState }) => {
+            await expect(
+              controller.addNewKeyring(MockKeyring.type),
+            ).rejects.toThrow('You will never be able to persist me!');
+
+            expect(controller.state.keyrings).toHaveLength(1);
+            // getAccounts calls each keyring to get its accounts
+            // resulting in an additional '0x1234' account being added
+            // in case of test failure
+            expect(await controller.getAccounts()).toStrictEqual(
```

Comment:
Hmm, this behaviour is changed,we might need to use something else to assert what we want

Reply:
`exportAccount` might work, if not for the fact that there is another keyring created automatically as part of `withController` that has exactly the same address present in it (due to the mocks).

Maybe we could use `exportAccount` but use the skip vault creation option of `withController`?

Reply:
`exportAccount` is perfect for the job, because we can control the address included in the mock keyring:
```typescript
stubKeyringClassWithAccount(MockKeyring, mockAddress);
// ...
await expect(
  controller.exportAccount(password, mockAddress),
).rejects.toThrow(
  'KeyringController - No keyring found. Error info: There are keyrings, but none match the address',
);
```